### PR TITLE
[feat] Added STE gradient estimator

### PR DESCRIPTION
**Pull request by shijianjian**
Added ` STEFunction` and `StraightThroughEstimator`

### CODE REVIEW

File: kornia/grad_estimator/ste.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        ctx, input: Tensor, output: Tensor,
+        grad_fn: Optional[Callable] = None
+    ) -> Tensor:
+        ctx.in_shape = input.shape
+        ctx.out_shape = output.shape
+        ctx.grad_fn = grad_fn
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output: Tensor) -> Tuple[Tensor, Tensor, None]:  # type:ignore
```
Comment (edgarriba) [inline kornia/grad_estimator/ste.py:52]: will this work with onnx/jit ? usually ignores doesn't support well tracing

**Review by edgarriba — APPROVED** [review]
LGTM. Do we have a direct use case // application ?

**Comment by shijianjian** [comment]
Not a full application yet. A straight forward use case can be:
```python
aug = nn.Sequential(
    K.RandomAffine(360),
    StraightThroughEstimator(K.RandomPosterize(3, p=1.), grad_fn=F.hardtanh)
    K.RandomRotate(360),
)

loss = aug(img).mean()
loss.backward()
```

Since the `K.RandomPosterize` is not differentiable, we can bypass its vanished gradients with this.

**Comment by edgarriba** [comment]
I see, this is a cool example to showcase ;)

File: kornia/grad_estimator/ste.py (hunk 1 above)
Comment (shijianjian) [reply kornia/grad_estimator/ste.py:53]: torch.Function does not support well with ONNX and JIT. Probably need to have a suite of C++ code for that.

https://pytorch.org/docs/1.10.0/onnx.html#torch-autograd-functions

https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html

File: kornia/grad_estimator/ste.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+from typing import Callable, Optional, Tuple
+
+import torch
```
Comment (shijianjian) [inline kornia/grad_estimator/ste.py:3]:
```suggestion
```
